contrib/node/src/python/pants/contrib/node/tasks/node_resolve.py

```diff
@@ -5,15 +5,65 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import os
+from hashlib import sha1
 
+from future.utils import PY3
 from pants.base.build_environment import get_buildroot
+from pants.base.fingerprint_strategy import DefaultFingerprintHashingMixin, FingerprintStrategy
 from pants.base.workunit import WorkUnitLabel
 from pants.build_graph.build_graph import sort_targets
 
 from pants.contrib.node.tasks.node_paths import NodePaths, NodePathsLocal
 from pants.contrib.node.tasks.node_task import NodeTask
 
 
+class NodeResolveFingerprintStrategy(DefaultFingerprintHashingMixin, FingerprintStrategy):
+  """
+  Fingerprint package lockfiles (e.g. package.json, yarn.lock...),
+  so that we don't automatically run this if none of those have changed.
+
+  We read every file and add its contents to the hash.
+  """
+
+  default_package_manager_lockfiles = {
+    'yarn': ['package.json', 'yarn.lock'],
+    'npm': ['package.json']
+  }
+
+  def validate_lockfile_in_sources(self, full_path, target_sources, target_address):
+    if full_path in target_sources:
+      return full_path
+    else:
+      raise ValueError("Trying to watch a file ({}) which is not in the sources of target {}".format(
```

Comment:
This should not fail if the file does not exist, because while it is recommended to commit lockfiles and include them in your sources, it is optional. I expect that the intent here is to assert that the file(s) are actually covered by fingerprinting if they exist, but your tests should cover that instead.

Having said that, I'm not sure that the per-target option is a good idea, as it will mean that by default it is required to have and commit a lockfile, and that's not currently a requirement.

Reply:
With the per-target configuration reverted, and this assertion removed, my shipit stands though.

Reply:
So, currently the plugin will throw if a lockfile is not in the `sources` field (particularly `package.json` and `yarn.lock` for `yarn` `node_module`s is what I've tested). However, this still means that we can remove the assertion, and re-implement the per-target thing if it becomes relevant.

Also, I agree that this still doesn't force users to _commit_ their lockfiles.